Library/Homebrew/livecheck/strategy/crate.rb

```diff
@@ -0,0 +1,115 @@
+# typed: true
+# frozen_string_literal: true
+
+module Homebrew
+  module Livecheck
+    module Strategy
+      # The {Crate} strategy identifies versions of a crates.io package by
+      # checking the information from the `versions` API endpoint.
+      #
+      # Crate URLs have the following format:
+      #   `https://static.crates.io/crates/example/example-1.2.3.crate`
+      #
+      # The default regex identifies versions like `1.2.3`/`v1.2.3` from the
+      # version `num` field. This is a common version format but a different
+      # regex can be provided in a `livecheck` block to override the default
+      # if a package uses a different format (e.g. `1.2.3d`, `1.2.3-4`, etc.).
+      #
+      # @api public
+      class Crate
+        # The default regex used to identify versions when a regex isn't
+        # provided.
+        DEFAULT_REGEX = /^v?(\d+(?:\.\d+)+)$/i
+
+        # The default `strategy` block used to extract version information when
+        # a `strategy` block isn't provided.
+        DEFAULT_BLOCK = proc do |json, regex|
+          json["versions"]&.map do |version|
+            next if version["yanked"] == true
+
+            match = version["num"]&.match(regex)
+            next if match.blank?
+
+            match[1]
+          end
+        end.freeze
+
+        # The `Regexp` used to determine if the strategy applies to the URL.
+        URL_MATCH_REGEX = %r{
+          ^https?://static\.crates\.io/crates
+          /(?<package>[^/]+) # The name of the package
+          /.+\.crate # The crate filename
+        }ix
```

Comment:
Would this prevent using the strategy for alternative/private registries?

Reply:
Yes but I'm happy to modify it if third-party registries can be checked in the same way and it's possible to create a regex that will work for everything (crates.io and third-parties). I'll look into how the [listed third-party registries](https://github.com/rust-lang/cargo/wiki/Third-party-registries) work in a bit.

Reply:
Looking through the [alternative registries RFC](https://github.com/rust-lang/rfcs/blob/master/text/2141-alternative-registries.md), [Registry Web API documentation](https://doc.rust-lang.org/cargo/reference/registry-web-api.html), and the code/documentation of some open source registries, it appears that the API for third-party registries isn't as complete if they're using something other than the crates.io codebase. Namely, the `/api/v1/crates/{crate}/versions` endpoint isn't in the Web API documentation and the limited number of open source registries that I looked at only provided a `max_version` value from `/api/v1/crates/{crate}` without including the `versions` array (like crates.io).

`max_version` may or may not give us the highest version but I'm not sure if that factors in yanked status, how it would work if a crate publishes unstable versions like `2.0.0-beta1` alongside stable versions like `1.2.3`, etc. Checking all the versions (and feeding them to a `strategy` block) seems like a more robust approach, so I don't think we should naively check `max_version` just to support third-party registries (not without confirming the behavior in the aforementioned scenarios, at least).

However, Markus's comment about a generic `Crate` name being more suitable if we support third-party registries makes sense to me, so I'm leaning in that direction again. Unless we come up with a clear plan of how we can support third-party registries, I think we should just put it off until we have an example we in homebrew/core to work with. This would also give the alternative registry implementations more time to mature and their API response format may be more in line with crates.io at that time.